### PR TITLE
IPC4: fix ipc_send_queued_msg() logic

### PR DESCRIPTION
**Pull request by kv2019i**
Fix issues with ipc_send_queued_msg() both in generic IPC4 code, as well as a seperate problem in Zephyr SOF builds.

FYI @bardliao  @abonislawski @marcinszkudlinski

### CODE REVIEW

**Comment by lgirdwood** [comment]
@mwasko fyi - pls review

**Review by lgirdwood — APPROVED** [review]
@kv2019i fix for v2.3 ? or IPC4 only

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> @kv2019i fix for v2.3 ? or IPC4 only

The other commit could hit Zephyr+IPC3 combo if you have a source of frequency notifications on Zephyr side (e.g. DMA trace, IPC stream position updates enabled or something similar). So potentially for 2.3 as well.

**Comment by kv2019i** [comment]
V2 uploaded:
 - one DEBUG trace left accidentally in the code, now removed
 - no other changes

**Comment by kv2019i** [comment]
V3 update:
- improved the commit message

**Comment by kv2019i** [comment]
V4 update:
- the ifdefs for ZEPHYR become too ugly, so I now modified implementation for both XTOS and Zephyr and move it up to IPC layer
- the EDF scheduler usage in earlier versions was misusing the interface (start field should be ignored by EDF scheduler but it's not on Zephyr now), so drop this and use a regular LL task instead

**Comment by kv2019i** [comment]
V5 update:
 - there is some unfortunate onion-peeling effect here
 - now that TX queueing works (and is automatic if EBUSY is seen), this uncovers a bug in IPC4 logging where a single notify message object was not properly protected for concurrent access

File: src/ipc/ipc-common.c
Comment (kv2019i) [reply src/ipc/ipc-common.c:214]: @serhiy-katsyuba-intel  This is outside my PR (and not my code), but I was wondering about just the same. In the end it does seem to be correct as the list_is_empty check on L207 is done on a different list object. E.g. if the msg's list entry is points to itself (e.g. is empty), then that message has not been added to the IPC protocol stacks list of pending messages. If it's non-empty, then it's already queued.

Comment (serhiy-katsyuba-intel) [reply src/ipc/ipc-common.c:214]: Yes, you are right, the code is correct. That item vs list distinguishing is not obvious at first look.

**Comment by kv2019i** [comment]
V6 update:
 - back to ready-for-review
 - one bugfix to the re-schedule logic

File: src/ipc/ipc4/handler.c
Comment (ujfalusi) [inline src/ipc/ipc4/handler.c:1037]: At the moment this should be OK as the only notification provider is mtrace, but in future if we have other notification sources then either each will have it's own static struct or we need to dynamically allocate notification messages..

**Review by ujfalusi — APPROVED** [review]

**Comment by kv2019i** [comment]
V7 update:
- this adds two fixes to mtrace logging that address a problem of high rate buffer-status notifies when mtrace is enabled but no mtrace-reader is running during use-cases

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
> @lgirdwood The first patch affects XTOS builds (moves IPC late sending from OS main loop to LL scheduler) and it is possible this is now failing cases (e.g. the Intel CI builds).

Zephyr build was failing on the zephyr.h issues, fix now merged but it looks like build still problematic. Maybe needs a rebase and force push ?

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> Zephyr build was failing on the zephyr.h issues, fix now merged but it looks like build still problematic. Maybe needs a rebase and force push ?

I think you need to ignore these GH tests that test with "main" of Zephyr. These are "informational" (for us), but it's is normal and expected that upstream Zephyr has merged patches that does not build with current SOF main. I.e. these checks will pass again when we make a SOF PR that updates the manifest (and potential local changes to SOF).

The IPC3 failure is more worrying. There seems to be a IPC panic in a multicore test.

**Comment by kv2019i** [comment]
Hmm, let me fork out the changes that do not have any impact to IPC3/XTOS.

File: src/ipc/ipc4/logging.c
Comment (lyakh) [inline src/ipc/ipc4/logging.c:80]: here updating `mtrace_notify_last_sent` is intended?

Comment (kv2019i) [reply src/ipc/ipc4/logging.c:80]: Damn, rebasing brought it back. Will address.

File: src/ipc/ipc-common.c
Comment (kv2019i) [reply src/ipc/ipc-common.c:226]: Hmm, indeed that seems to be safe (or at least done elsewhere in the code. I'll move it.

File: src/ipc/ipc-common.c
Comment (kv2019i) [reply src/ipc/ipc-common.c:262]: Yup, I dropped that approached and used LL task instead. We should not rely on the "task start" field in EDF scheduling (it's a feature of the Zephyr SOF EDF scheduler implementation, not part of the interface).

**Comment by kv2019i** [comment]
I'll close this to not waste reviewers time. There's a repeating failure in XTOS/IPC3 tests with this series and this probably relates to the first patch that also touches the XTOS primary core main loop (moving the IPC flushing from OS main loop to a normal LL scheduler task). Until I figure this out, we can't merge this. The other fixes are important though, so let's get them merged at a faster speed. -> #6260 

Sorry for the churn all reviewers.

**Comment by kv2019i** [comment]
V8 update:
 - reopened the PR
 - 4/5 commits merged in another PR, now only one commit left in this PR
 - marking as draft until the XTOS/IPC3 failure in CI is rootcaused

**Comment by kv2019i** [comment]
Still failing in the multicore IPC3/XTOS tests.

**Comment by kv2019i** [comment]
It turns out this change is overlapping with c194125b83c21d82243bc9eaa18701d87deba468 @marc-hb  did last year. So IOW, Zephyr+IPC3 combination is good after all (which makes a lot more sense). Combined with the errors found on XTOS by CI, I'll close this again.

**Pull request closed without merging**
